### PR TITLE
Fix markdown math brackets render problem

### LINKED ISSUE #31371: Markdown doesn't render math within brackets
### Description

It seems markdown doesn't render math within brackets.

For example, on Github the markdown
```markdown
$i = \sqrt{-1}$ ($i = \sqrt{-1}$)
```
renders as

$i = \sqrt{-1}$ ($i = \sqrt{-1}$)

However, on Gitea it renders differently
9ef3e3b7-c81f-428b-9b0c-f097b2558e39

### Gitea Version

1.23.0+dev-218-g90bcdf9829

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Reproduced at [https://demo.gitea.com/](https://demo.gitea.com/lns/test)

### Database

None

### PR DESCRIPTION
**Pull request by charles7668**
close #31371

compare

original text:
```text
$i = \sqrt{-1}$ ($i = \sqrt{-1}$)

$i = \sqrt{-1}$ [$i = \sqrt{-1}$]

$i = \sqrt{-1}$ {$i = \sqrt{-1}$}

$abcd

(this is a test)

[$test]

$test$

$a$ ($b$) [$c$] {$d$}
```

before fix:
image

after fix:
image

### CODE REVIEW

File: modules/markup/markdown/math/inline_parser.go [hunk 1]
```diff
@@ -45,6 +45,10 @@ func isPunctuation(b byte) bool {
 	return b == '.' || b == '!' || b == '?' || b == ',' || b == ';' || b == ':'
 }
 
+func isBracket(b byte) bool {
+	return b == '(' || b == ')' || b == '[' || b == ']' || b == '{' || b == '}'
```
Comment (wxiaoguang) [inline modules/markup/markdown/math/inline_parser.go]: Overall LGTM.

GitHub only supports `($ $)`, so I would suggest to follow its behavior until there is a clear requirement for other cases.

image

File: modules/markup/markdown/math/inline_parser.go [hunk 2]
```diff
@@ -84,7 +88,7 @@ func (parser *inlineParser) Parse(parent ast.Node, block text.Reader, pc parser.
 			break
 		}
 		suceedingCharacter := line[pos]
-		if !isPunctuation(suceedingCharacter) && !(suceedingCharacter == ' ') {
+		if !isPunctuation(suceedingCharacter) && !(suceedingCharacter == ' ') && !isBracket(suceedingCharacter) {
```
Comment (wxiaoguang) [inline modules/markup/markdown/math/inline_parser.go:91]: Maybe it should only check `)` because `suceedingCharacter` can't be `(`?

Comment (charles7668) [reply modules/markup/markdown/math/inline_parser.go:91]: Just consider that some weird patterns might exist, so I added `(`.

Maybe checking for `)` is enough.

Comment (wxiaoguang) [reply modules/markup/markdown/math/inline_parser.go:91]: Thank you! Let's keep it simple to avoid side effects, then add more supports one by one 😁

**Comment by wxiaoguang** [comment]
Thank you for the fix!

ps: I will edit the PR title to remove the `(# xxx)` part: that part is only use for backporting in Gitea.

**Review by wxiaoguang — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Pull request merged**